radiation/TwoStreamMLPEMod.F90

```diff
@@ -0,0 +1,1723 @@
+  ! are divisions of horizontal space, ie literal columns of space. The current
+  ! implementation limits this space to media that has uniform scattering coefficients.
+  ! So there could not be different PFTs in the same column, because they would undoubtedly
+  ! have different joint scattering coefficients at different height levels in
+  ! the column.  Therefore, every column is connected with a PFT.
+
+
+  type, public :: twostream_type
+
+     type(scelg_type), pointer :: scelg(:,:)    ! array of scattering elements (layer, column)
+                                                ! can be sparse, will only solve indices up to
+                                                ! n_lyr,n_col(n_lyr). This is for band (wavelength)
+                                                ! independent information
+
+     type(band_type), pointer  :: band(:)       ! Holds scattering coefficients for each band
+                                                ! vis,nir,etc (nothing that emits though, no thermal)
+
+     integer                   :: n_bands       ! number of bands (allocation size of band(:))
+     integer                   :: n_lyr         ! number of (vertical) scattering element layers
+     integer, allocatable      :: n_col(:)      ! number of (horizontal) scattering element columns per layer
+     integer                   :: n_scel        ! total number of scattering elements
+     logical                   :: force_prep    ! Some coefficients are only updated
+                                                ! when the canopy composition changes, ie
+                                                ! changes in leaf, stem or snow structure.
+                                                ! If so, this sets to true, signalling that diffuse
+                                                ! scattering coefficients should be updated.
+                                                ! Otherwise, we only updated zenith dependent
+                                                ! parameters on short sub-daily timesteps
+     real(r8)                  :: frac_snow     ! Current mean snow-fraction of the canopy
+     real(r8)                  :: frac_snow_old ! Previous mean snow-fraction of the canopy
+     real(r8)                  :: cosz          ! Current cosine of the zenith angle
+     
+   contains
+
+     procedure :: ZenithPrep     ! Update coefficients as zenith changes
+     procedure :: CanopyPrep     ! Update coefficients as canopy changes
+     procedure :: Solve          ! Perform the scattering solution
+     procedure :: Dump           ! Dump out (print out) the site of interest
+     procedure :: GetNSCel
+     procedure :: AllocInitTwoStream
+     procedure :: DeallocTwoStream
+     
+     procedure :: GetRdUp
+     procedure :: GetRdDn
+     procedure :: GetRb
+     procedure :: GetAbsRad
+     
+
+  end type twostream_type
+
+  public :: ParamPrep
+  public :: AllocateRadParams
+  public :: TwoStreamLogInit
```

Comment:
Why is it necessary to have a specific log initialization for 2stream (instead of using the `fates_log` global?  Is it supportive of unit testing?

Reply:
yes, it made unit testing a little easier